tests/Temporalio.Tests/Worker/ActivityWorkerTests.cs

```diff
@@ -0,0 +1,718 @@
+        [Activity]
+        async Task WaitUntilCancelledAsync()
+        {
+            activityWaiting.SetResult();
+            try
+            {
+                await Task.Delay(Timeout.Infinite, ActivityContext.Current.CancellationToken);
+            }
+            catch (TaskCanceledException) when (
+                ActivityContext.Current.CancelReason == ActivityCancelReason.WorkerShutdown)
+            {
+                workerShutdown = true;
+                throw;
+            }
+        }
+
+        var taskQueue = $"tq-{Guid.NewGuid()}";
+        using var worker = new TemporalWorker(Client, new()
+        {
+            TaskQueue = taskQueue,
+            Activities = { WaitUntilCancelledAsync },
+        });
+        // Overwrite bridge worker with one we can inject failures on
+        var bridgeWorker = new ManualPollCompletionBridgeWorker(worker.BridgeWorker);
+        worker.BridgeWorker = bridgeWorker;
+
+        // Run the worker
+        WorkflowHandle? handle = null;
+        var wErr = await Assert.ThrowsAsync<InvalidOperationException>(
+            async () => await worker.ExecuteAsync(async () =>
+        {
+            // Start the workflow
+            handle = await Env.Client.StartWorkflowAsync(
+                IKitchenSinkWorkflow.Ref.RunAsync,
+                new KSWorkflowParams(new KSAction(ExecuteActivity: new(
+                    Name: "WaitUntilCancelled",
+                    TaskQueue: taskQueue))),
+                new(id: $"workflow-{Guid.NewGuid()}", taskQueue: Env.KitchenSinkWorkerTaskQueue));
+
+            // Wait for activity to report waiting, then send a poll failure
+            await activityWaiting.Task;
+            bridgeWorker.PollActivityCompletion.SetException(
+                new InvalidOperationException("Oh no"));
+            await Task.Delay(Timeout.Infinite);
+        }));
+        Assert.Equal("Oh no", wErr.Message);
+
+        // Confirm workflow cancelled
+        var wfErr = await Assert.ThrowsAsync<WorkflowFailedException>(
+            () => handle!.GetResultAsync());
+        var actErr = Assert.IsType<ActivityFailureException>(wfErr.InnerException);
+        var appErr = Assert.IsType<ApplicationFailureException>(actErr.InnerException);
+        Assert.Equal("TaskCanceledException", appErr.ErrorType);
+        Assert.True(workerShutdown);
+    }
+
+    [Fact]
+    public void New_DuplicateActivityNames_Throws()
+    {
+        [Activity("some-activity")]
+        static string SomeActivity1() => string.Empty;
+        [Activity("some-activity")]
+        static string SomeActivity2() => string.Empty;
+        var err = Assert.Throws<ArgumentException>(() =>
+        {
+            using var worker = new TemporalWorker(Client, new()
+            {
+                TaskQueue = $"tq-{Guid.NewGuid()}",
+                Activities = { SomeActivity1, SomeActivity2 },
+            });
+        });
+        Assert.Contains("same key has already been added", err.Message);
```

Comment:
Not sure what the entire message looks like but can we check that it mentions that it's a duplicate activity registration error and it says which activities are conflicting?

Reply:
I admit being lazy and just letting https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.todictionary?view=net-7.0 throw which does say what is duplicated but not that it's activities. Can improve.

Reply:
Fixed